### DISCUSSION TITLE
Measure MMC performances

### DISCUSSION THREAD

**Discussion by blemouzy** [zephyrproject-rtos/zephyr#61943] [Q&A] [answered]
Hello,

I want to get an estimation of MMC performances on a custom i.MX RT1052 board.

On Linux, I use:
- `dd` to write data with flag to sync data
- `echo 3 > /proc/sys/vm/drop_caches` to drop caches
- `dd` to read data

Is there anything which I take care about (like dropping caches) with Zephyr using `mmc_write_blocks` / `mmc_read_blocks` functions?

**Comment by DerekSnell** [comment, marked as answer]
Hi @blemouzy ,

This is some feedback I received for your question:
Since Zephyr has no software caching in the disk subsystem, there shouldn’t be anything special needed beyond `mmc_write_blocks` / `mmc_read_blocks`. It’s worth noting we have a [test for disk performance](https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/disk/disk_performance) in Zephyr. We do enable the on-device eMMC cache during [disk init](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/sd/mmc.c#L197), but based on what was found online regarding `/proc/sys/vm/drop_caches`, this command will only drop page caches as well as cached dentries and inodes. So I think that only software-based caching is affected, not on-device eMMC cache.

**Comment by blemouzy** [comment]
Hi @DerekSnell ,

> Since Zephyr has no software caching in the disk subsystem, there shouldn’t be anything special needed beyond `mmc_write_blocks` / `mmc_read_blocks`. 
> [...]
> So I think that only software-based caching is affected, not on-device eMMC cache.

Thanks for these information.

> It’s worth noting we have a [test for disk performance](https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/disk/disk_performance) in Zephyr.

I tried to use it on my board but all `disk_` functions failed.
Replace them with `mmc_` equivalent was super quick and the result is satisfactory:

```
I: Device sai@40384000 initialized
W: No power control GPIO defined. Without power control,
the SD card may fail to communicate with the host
*** Booting Zephyr OS build v3.4.0 ***
Running TESTSUITE mmc_performance
===================================================================
I: Card does not support CMD8, assuming legacy card
I: Card switched to 1.8V signaling
I: CID decoding not supported for MMC
I: Card block count is 15269888, block size is 512
MMC reports 15269888 sectors
MMC reports sector size 512
START - test_random_read
512 Byte IOPS over 16384 random reads: 5257 IOPS
 PASS - test_random_read in 3.135 seconds
===================================================================
START - test_random_write
512 Byte IOPS over 16384 random writes: 2847 IOPS
 PASS - test_random_write in 1.861 seconds
===================================================================
START - test_sequential_read
Average read speed over one sector: 152 KiB/s
Average read speed over 16384 sectors: 40960 KiB/s
 PASS - test_sequential_read in 1.874 seconds
===================================================================
START - test_sequential_write
Average write speed over one sector: 8270 KiB/s
Average write speed over 16384 sectors: 24576 KiB/s
 PASS - test_sequential_write in 3.841 seconds
===================================================================
TESTSUITE mmc_performance succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [mmc_performance]: pass = 4, fail = 0, skip = 0, total = 4 duration = 10.711 seconds
 - PASS - [mmc_performance.test_random_read] duration = 3.135 seconds
 - PASS - [mmc_performance.test_random_write] duration = 1.861 seconds
 - PASS - [mmc_performance.test_sequential_read] duration = 1.874 seconds
 - PASS - [mmc_performance.test_sequential_write] duration = 3.841 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION SUCCESSFUL
```

Thanks!
